File: src/test/java/com/youngjun/auth/core/api/controller/v1/TokenControllerTest.java
package com.youngjun.auth.core.api.controller.v1;

import com.youngjun.auth.core.api.controller.v1.request.ReissueTokenRequestDto;
import com.youngjun.auth.core.api.support.RestDocsTest;
import com.youngjun.auth.core.api.support.RestDocsUtils;
import com.youngjun.auth.core.domain.token.TokenPair;
import com.youngjun.auth.core.domain.token.TokenService;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullSource;
import org.junit.jupiter.params.provider.ValueSource;
import org.mockito.ArgumentMatchers;
import org.mockito.BDDMockito;
import org.mockito.Mockito;
import org.springframework.http.MediaType;
import org.springframework.restdocs.mockmvc.MockMvcRestDocumentation;
import org.springframework.restdocs.payload.JsonFieldType;
import org.springframework.restdocs.payload.PayloadDocumentation;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultHandlers;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;

import java.nio.charset.StandardCharsets;

class TokenControllerTest extends RestDocsTest {

    private static final String ACCESS_TOKEN = "${ACCESS_TOKEN}";

    private static final Long ACCESS_TOKEN_EXPIRES_IN = 1L;

    private static final String REFRESH_TOKEN = "${REFRESH_TOKEN}";

    private static final Long REFRESH_TOKEN_EXPIRES_IN = 1L;

    private final TokenService tokenService = Mockito.mock(TokenService.class);

    @Override
    protected Object initController() {
        return new TokenController(tokenService);
    }

    @DisplayName("재발급 성공")
    @Test
    void reissue() throws Exception {
        ReissueTokenRequestDto request = new ReissueTokenRequestDto("${REFRESH_TOKEN}");

        BDDMockito.given(tokenService.reissue(ArgumentMatchers.anyString())).willReturn(buildTokenPair());

        mockMvc
                .perform(MockMvcRequestBuilders.post("/auth/token")
                        .content(objectMapper.writeValueAsString(request))
                        .contentType(MediaType.APPLICATION_JSON)
                        .characterEncoding(StandardCharsets.UTF_8))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.status().isOk())
                .andDo(MockMvcRestDocumentation.document("token", RestDocsUtils.requestPreprocessor(),
                        RestDocsUtils.responsePreprocessor(),
                        PayloadDocumentation.requestFields(PayloadDocumentation.fieldWithPath("refreshToken")
                                .type(JsonFieldType.STRING)
                                .description("refreshToken")),
                        PayloadDocumentation.responseFields(
                                PayloadDocumentation.fieldWithPath("status")
                                        .type(JsonFieldType.STRING)
                                        .description("status"),
                                PayloadDocumentation.fieldWithPath("data").type(JsonFieldType.OBJECT).description("data"),
                                PayloadDocumentation.fieldWithPath("data.accessToken")
                                        .type(JsonFieldType.STRING)
                                        .description("accessToken"),
                                PayloadDocumentation.fieldWithPath("data.accessTokenExpiresIn")
                                        .type(JsonFieldType.NUMBER)
                                        .description("accessToken 만료 시간, UNIX 타임스탬프(Timestamp)"),
                                PayloadDocumentation.fieldWithPath("data.refreshToken")
                                        .type(JsonFieldType.STRING)
                                        .description("refreshToken"),
                                PayloadDocumentation.fieldWithPath("data.refreshTokenExpiresIn")
                                        .type(JsonFieldType.NUMBER)
                                        .description("refreshToken 만료 시간, UNIX 타임스탬프(Timestamp)"),
                                PayloadDocumentation.fieldWithPath("error")
                                        .type(JsonFieldType.NULL)
                                        .description("error")
                                        .ignored())));
    }

    @DisplayName("유효하지 않은 refresh token 으로 재발급 시 실패한다.")
    @ParameterizedTest
    @NullSource
    @ValueSource(strings = {"", // Empty String
            " ", // Only whitespace
    })
    void reissueWithInvalidRefreshToken(String invalidRefreshToken) throws Exception {
        ReissueTokenRequestDto request = new ReissueTokenRequestDto(invalidRefreshToken);

        BDDMockito.given(tokenService.reissue(ArgumentMatchers.anyString())).willReturn(null);

        mockMvc
                .perform(MockMvcRequestBuilders.post("/auth/token")
                        .content(objectMapper.writeValueAsString(request))
                        .contentType(MediaType.APPLICATION_JSON)
                        .characterEncoding(StandardCharsets.UTF_8))
                .andDo(MockMvcResultHandlers.print())
                .andExpect(MockMvcResultMatchers.status().isBadRequest());
    }

    private TokenPair buildTokenPair() {
        return new TokenPair(1L, ACCESS_TOKEN, ACCESS_TOKEN_EXPIRES_IN, REFRESH_TOKEN, REFRESH_TOKEN_EXPIRES_IN);
    }

}

File: src/test/kotlin/com/youngjun/auth/core/api/controller/v1/TokenControllerTest.kt
package com.youngjun.auth.core.api.controller.v1

import com.youngjun.auth.core.api.application.TokenService
import com.youngjun.auth.core.api.controller.v1.request.ReissueTokenRequest
import com.youngjun.auth.core.api.support.RestDocsTest
import com.youngjun.auth.core.api.support.description
import com.youngjun.auth.core.api.support.ignored
import com.youngjun.auth.core.api.support.optional
import com.youngjun.auth.core.api.support.type
import com.youngjun.auth.core.domain.token.RefreshToken
import com.youngjun.auth.core.domain.token.TokenPairBuilder
import io.mockk.every
import io.mockk.mockk
import io.restassured.http.ContentType
import io.restassured.module.mockmvc.RestAssuredMockMvc.given
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.document
import org.springframework.restdocs.payload.JsonFieldType.NULL
import org.springframework.restdocs.payload.JsonFieldType.NUMBER
import org.springframework.restdocs.payload.JsonFieldType.OBJECT
import org.springframework.restdocs.payload.JsonFieldType.STRING
import org.springframework.restdocs.payload.PayloadDocumentation.requestFields
import org.springframework.restdocs.payload.PayloadDocumentation.responseFields

class TokenControllerTest : RestDocsTest() {
    private lateinit var tokenService: TokenService

    @BeforeEach
    fun setUp() {
        tokenService = mockk()
        val controller = TokenController(tokenService)
        setMockMvc(controller)
    }

    @Test
    fun `재발급 성공`() {
        val value = "refreshToken"
        every { tokenService.reissue(RefreshToken(value)) } returns TokenPairBuilder().build()

        given()
            .log()
            .all()
            .contentType(ContentType.JSON)
            .body(ReissueTokenRequest(value))
            .post("/auth/token")
            .then()
            .log()
            .all()
            .apply(
                document(
                    "token",
                    requestFields(
                        "refreshToken" type STRING description "refreshToken",
                    ),
                    responseFields(
                        "status" type STRING description "status",
                        "data" type OBJECT description "data",
                        "data.accessToken" type STRING description "accessToken",
                        "data.accessTokenExpiresIn" type NUMBER description "accessToken 만료 시간, UNIX 타임스탬프(Timestamp)",
                        "data.refreshToken" type STRING description "refreshToken" optional true,
                        "data.refreshTokenExpiresIn" type NUMBER description "refreshToken 만료 시간, UNIX 타임스탬프(Timestamp)" optional true,
                        "error" type NULL ignored true,
                    ),
                ),
            )
    }
}
